File: backend/app/schemas/review.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel


class TaskReviewItem(BaseModel):
    id: UUID
    title: str
    description: str | None = None
    due_date: datetime | None = None
    created_at: datetime | None = None


class ProjectReviewItem(BaseModel):
    id: UUID
    name: str
    description: str | None = None
    has_next_action: bool
    days_without_next: int = 0
    next_actions: list[TaskReviewItem] = []
    available_tasks: list[TaskReviewItem] = []


class ReviewAlert(BaseModel):
    severity: Literal['red', 'yellow']
    message: str
    project_id: str | None = None


class PreviousSnapshot(BaseModel):
    created_at: datetime
    inbox_count: int
    overdue_count: int
    done_count: int
    stale_count: int
    projects_without_next: int
    health_status: str


class ReviewSummaryResponse(BaseModel):
    inbox_count: int
    overdue_count: int
    done_this_week: int
    stale_count: int
    someday_count: int = 0
    projects_without_next: int
    health_status: Literal['ok', 'attention', 'problems']
    last_review_date: datetime | None = None
    review_count: int
    review_frequency_days: int = 7
    week_activity: dict[str, int] = {}
    alerts: list[ReviewAlert] = []
    previous_snapshot: PreviousSnapshot | None = None


class OverdueTaskItem(BaseModel):
    id: UUID
    title: str
    description: str | None = None
    due_date: datetime | None = None
    gtd_status: str
    project_name: str | None = None


class ReviewStatusResponse(BaseModel):
    inbox_count: int
    inbox_tasks: list[TaskReviewItem]
    overdue_tasks: list[OverdueTaskItem] = []
    active_projects: list[ProjectReviewItem]
    someday_tasks: list[TaskReviewItem]
    last_review_date: datetime | None = None
    review_count: int


class ReviewCompleteResponse(BaseModel):
    success: bool
    review_count: int
    completed_at: str
    snapshot_health: str | None = None

File: frontend/src/api/review.ts
import { httpClient } from './httpClient'

export interface TaskReviewItem {
  id: string
  title: string
  description: string | null
  due_date: string | null
  created_at: string | null
}

export interface ProjectReviewItem {
  id: string
  name: string
  description: string | null
  has_next_action: boolean
  days_without_next: number
  next_actions: TaskReviewItem[]
  available_tasks: TaskReviewItem[]
}

export interface ReviewAlert {
  severity: 'red' | 'yellow'
  message: string
  project_id: string | null
}

export interface ReviewSummary {
  inbox_count: number
  overdue_count: number
  done_this_week: number
  stale_count: number
  someday_count: number
  projects_without_next: number
  health_status: 'ok' | 'attention' | 'problems'
  last_review_date: string | null
  review_count: number
  review_frequency_days: number
  week_activity: Record<string, number>
  alerts: ReviewAlert[]
}

export interface OverdueTaskItem {
  id: string
  title: string
  description: string | null
  due_date: string | null
  gtd_status: string
  project_name: string | null
}

export interface ReviewStatus {
  inbox_count: number
  inbox_tasks: TaskReviewItem[]
  overdue_tasks: OverdueTaskItem[]
  active_projects: ProjectReviewItem[]
  someday_tasks: TaskReviewItem[]
  last_review_date: string | null
  review_count: number
}

export interface ReviewCompleteRequest {
  inbox_processed: number
  next_actions_added: number
  someday_activated: number
}

export interface ReviewCompleteResponse {
  success: boolean
  review_count: number
  completed_at: string
  snapshot_health: string | null
}

export const reviewApi = {
  getSummary: async (): Promise<ReviewSummary> => {
    const response = await httpClient.get<ReviewSummary>('/review/summary')
    return response.data
  },

  getStatus: async (): Promise<ReviewStatus> => {
    const response = await httpClient.get<ReviewStatus>('/review/status')
    return response.data
  },

  complete: async (stats?: ReviewCompleteRequest): Promise<ReviewCompleteResponse> => {
    const response = await httpClient.post<ReviewCompleteResponse>('/review/complete', stats)
    return response.data
  },
}
